### PR TITLE
Spell crafting

### PR DESCRIPTION
**Pull request by Sanko-Kallig**
There are still some TODO's left open, but these have been formed into their own tasks on Trello

### CODE REVIEW

File: RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        public List<Fragment> Bin { get; private set; }
+        #endregion
+        #region Private
+        /// <summary>
+        /// Datasets after splitting the bin into managable pieces
+        /// </summary>
+        private List<DataSet> DataSets;
+
+        /// <summary>
+        ///
```
Comment (Sven-dev) [inline RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs]: Empty summary

File: RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            for (int i = 0; i < this.setSize; i++)
+            {
+                if (DataSets[i].IsSolved)
+                    solvedCount++;
+            }
+            if (solvedCount == setSize)
+            {
+                return true;
+            }
+            else
```
Comment (Sven-dev) [inline RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs]: else statement not needed

File: RPG Wizardry/Assets/Scripts/ResearchData/DataManager.cs [hunk 3]
```diff
@@ -0,0 +1,113 @@
+﻿using nl.SWEG.RPGWizardry.Player.Inventory;
```
Comment (Sven-dev) [inline RPG Wizardry/Assets/Scripts/ResearchData/DataManager.cs]: Add summaries to this script

File: RPG Wizardry/Assets/Scripts/ResearchData/Fragment.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace nl.SWEG.RPGWizardry.ResearchData
+{
+
+    public class Fragment
+    {
+        #region Methods
+        #region Public
+        /// <summary>
+        ///
```
Comment (Sven-dev) [inline RPG Wizardry/Assets/Scripts/ResearchData/Fragment.cs:15]: Empty summary

**Review by Sven-dev — APPROVED** [review]

File: RPG Wizardry/Assets/Art/Sprites/UI/Element Icons/desktop.ini [hunk 5]
```diff
@@ -0,0 +1,5 @@
+[.ShellClassInfo]
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Art/Sprites/UI/Element Icons/desktop.ini]: Add to GitIgnore

File: RPG Wizardry/Assets/Scripts/Player/PlayerInput/SpellCraftingInput.cs [hunk 6]
```diff
@@ -57,8 +57,12 @@ private void MouseInputs()
 
                 if (results.Count > 0)
                 {
-                    draggedImage = results[0].gameObject.GetComponent<Image>();
-                    mousePos = Input.mousePosition;
+                    if(results[0].gameObject.tag == "Bar")
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/Player/PlayerInput/SpellCraftingInput.cs]: Use a [TagSelector]-annotation and a Serialized String (See ACollectable. Annotation can be put on either string or string[]) for tags, to ensure they use existing and valid tags

File: RPG Wizardry/Assets/Scripts/ResearchData/ControlFragment.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+{
+
+    public class ControlFragment : Fragment
+    {
+        #region Methods
+        #region Public
+        /// <summary>
+        /// The position which is the solution to this control fragment.
+        /// </summary>
+        public float CorrectPosition { get; set; }
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/ResearchData/ControlFragment.cs]: Don't add get/set if both are public.

File: RPG Wizardry/Assets/Scripts/ResearchData/ControlFragment.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        /// <summary>
+        /// Used to determine if this fragment is solved or not.
+        /// </summary>
+        public bool Solved { get; private set; }
+        /// <summary>
+        /// the offset from the correct position which will count as a valid solution
+        /// </summary>
+        #endregion
+        #region Private
+        private float range;
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/ResearchData/ControlFragment.cs:25]: Add summaries

File: RPG Wizardry/Assets/Scripts/ResearchData/ControlFragment.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+            this.range = range;
+        }
+        #endregion
+        #region Check
+        /// <summary>
+        /// Checks the current position of the control fragment versus the correct position + offsets to see if this fragment has been solved.
+        /// </summary>
+        public override void CheckIfSolved()
+        {
+            if(ImageTransform.localPosition.y <= CorrectPosition + range && ImageTransform.localPosition.y >= CorrectPosition - range)
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/ResearchData/ControlFragment.cs:40]: Set solved to the result of the if? (edge-case: Solve, then change)

File: RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        /// 
+        /// </summary>
+        private bool IsSolved;
+        /// <summary>
+        /// Definition of how bin will be divided into DataSets
+        /// </summary>
+        private int setSize;
+        #endregion
+
+        public DataBin(List<Fragment> bin, int setSize)
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs:27]: Cleanup (Summaries, Region)

File: RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs (hunk 2 above)
Comment (frankvHoof93) [reply RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs]: return solvedCount == setSize;

Comment (frankvHoof93) [reply RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs]: Alternatively, use Linq: (!DataSets.Any(set -> !set.IsSolved))

File: RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+            if (solvedCount == setSize)
+            {
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+        public DataSet UnsolvedDataSet()
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs]: Rename to FirstUnsolvedDataSet, as it only returns the first unsolved set found

File: RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+            }
+
+        }
+        #endregion
+    }
+    #region Utility
+    /// <summary>
+    /// Helper methods for the lists.
+    /// </summary>
+    public static class ListExtensions
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs]: Move to Utils-NameSpace (e.g. Utils.Functions)

File: RPG Wizardry/Assets/Scripts/ResearchData/DataManager.cs [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace nl.SWEG.RPGWizardry.ResearchData
+{
+    public class DataManager : MonoBehaviour
+    {
+        public PlayerInventory Inventory;
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/ResearchData/DataManager.cs]: Is Public set required on these variables? (Inventory can be found via PlayerManager)

File: RPG Wizardry/Assets/Scripts/ResearchData/DataManager.cs [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+        /// <summary>
+        /// Perform start when page is enabled again.
+        /// </summary>
+        private void OnEnable()
+        {
+            Start();
+        }
+
+        // Update is called once per frame
+        void Update()
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/ResearchData/DataManager.cs:62]: Accessibility-Modifier

File: RPG Wizardry/Assets/Scripts/ResearchData/DataManager.cs [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+        private void ClearTexture(Texture2D tex)
+        {
+            for (int x = 0; x < tex.width; x++)
+                for (int y = 0; y < tex.height; y++)
+                    tex.SetPixel(x, y, Color.white);
+            tex.Apply();
+        }
+
+        public void ClearTexture()
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/ResearchData/DataManager.cs]: Rename to ClearTextures (plural)

File: RPG Wizardry/Assets/Scripts/ResearchData/DataSet.cs [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    public class DataSet
+    {
+        #region Methods
+        #region Public
+        /// <summary>
+        /// List of fragments for this dataset
+        /// </summary>
+        public List<Fragment> Fragments { get; set; }
+
+        public bool IsSolved { get; set; }
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/ResearchData/DataSet.cs]: Public Set required? (Don't use get/set if both are public)

File: RPG Wizardry/Assets/Scripts/ResearchData/DataSet.cs [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+            foreach (ControlFragment control in Fragments.OfType<ControlFragment>())
+            {
+                control.CheckIfSolved();
+                if (control.Solved)
+                {
+                    SolvedCount++;
+                }
+                    
+            }
+            if(Fragments.OfType<ControlFragment>().Count() == SolvedCount)
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/ResearchData/DataSet.cs]: return .Count() == SolvedCount

Store the OfType() to a local variable, so you don't have to call it twice

File: RPG Wizardry/Assets/Scripts/ResearchData/DataStub/DataStubBin.cs [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+using UnityEditor;
+using System.Collections.Generic;
+
+namespace nl.SWEG.RPGWizardry.ResearchData
+{
+    public class DataStubBin
+    {
+        public DataBin Bin = new DataBin(new List<Fragment>
+        {
+            new Fragment()
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/ResearchData/DataStub/DataStubBin.cs:11]: Is there a nicer way to do this? (e.g. JSON)

File: RPG Wizardry/Assets/Scripts/ResearchData/Fragment.cs [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+        /// Original position of the fragment in the fragment bin.
+        /// </summary>
+        public int OriginalRow { get; set; }
+        /// <summary>
+        /// UI Transform object needed to get the position.
+        /// </summary>
+        public Transform ImageTransform { get; set; }
+        #endregion
+
+        public Fragment()
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/ResearchData/Fragment.cs]: No need to add empty constructor

File: RPG Wizardry/Assets/Scripts/Sorcery/SpellPage.cs [hunk 20]
```diff
@@ -11,6 +11,11 @@ public class SpellPage
         /// Whether the Spell on this Page has been Unlocked
         /// </summary>
         public bool Unlocked { get; private set; }
+
+        /// <summary>
+        /// The dust cost for the spell to unlock.  
+        /// </summary>
+        public uint DustCost { get; private set; }
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/Sorcery/SpellPage.cs:19]: Should this be in SpellData-ScriptableObject? (I did that in my (open) branch)

File: RPG Wizardry/Assets/Scripts/Sorcery/SpellPageManager.cs [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+        [SerializeField]
+        private PlayerInventory inventory;
+        /// <summary>
+        /// Title of spell page
+        /// </summary>
+        [SerializeField]
+        private TextMeshProUGUI title;
+        [SerializeField]
+        private Button button;
+        void Start()
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/Sorcery/SpellPageManager.cs]: Add accessibility-modifier

File: RPG Wizardry/Assets/Scripts/Sorcery/SpellPageManager.cs [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+        /// <summary>
+        /// Spell targeted for unlocking
+        /// </summary>
+        [SerializeField]
+        private SpellPage selectedSpell;
+        /// <summary>
+        /// Player inventory  which contains current spell pages;
+        /// </summary>
+        [SerializeField]
+        private PlayerInventory inventory;
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/Sorcery/SpellPageManager.cs:20]: grab from PlayerManager instead

File: RPG Wizardry/Assets/Scripts/Sorcery/SpellPageManager.cs [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        /// <summary>
+        /// Unlocking of spell
+        /// </summary>
+        public void UnlockSpell()
+        {
+            if(inventory != null)
+            {
+                inventory.UnlockSpell(selectedSpell);
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/Sorcery/SpellPageManager.cs]: inventory?.UnlockSpell(selectedSpell) does a null-check for you ;)

File: RPG Wizardry/Assets/Scripts/Sorcery/SpellPageManager.cs [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+            }
+            
+        }
+        // Update is called once per frame
+        void Update()
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/Sorcery/SpellPageManager.cs]: Add accessibility-modifier

File: RPG Wizardry/Assets/Scripts/UI/Utils/Direction.cs [hunk 25]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
```
Comment (frankvHoof93) [inline RPG Wizardry/Assets/Scripts/UI/Utils/Direction.cs]: Why was this moved to UI?

**Review by frankvHoof93 — CHANGES_REQUESTED** [review]
Lot's of cleanup required

File: RPG Wizardry/Assets/Scripts/UI/Utils/Direction.cs (hunk 25 above)
Comment (Sanko-Kallig) [reply RPG Wizardry/Assets/Scripts/UI/Utils/Direction.cs]: duplicate that got created by a merge conflict apparently

File: RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs (hunk 1 above)
Comment (Sanko-Kallig) [reply RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs]: Done

File: RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs (hunk 2 above)
Comment (Sanko-Kallig) [reply RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs]: Also done

File: RPG Wizardry/Assets/Scripts/ResearchData/ControlFragment.cs (hunk 8 above)
Comment (Sanko-Kallig) [reply RPG Wizardry/Assets/Scripts/ResearchData/ControlFragment.cs:25]: Was already there somehow got split by the regions

**Review by frankvHoof93 — APPROVED** [review]

File: RPG Wizardry/Assets/Scripts/Sorcery/SpellPage.cs (hunk 20 above)
Comment (Sanko-Kallig) [reply RPG Wizardry/Assets/Scripts/Sorcery/SpellPage.cs:19]: Added a TODO for this in the code.

File: RPG Wizardry/Assets/Scripts/ResearchData/DataStub/DataStubBin.cs (hunk 18 above)
Comment (Sanko-Kallig) [reply RPG Wizardry/Assets/Scripts/ResearchData/DataStub/DataStubBin.cs:11]: Added a Trello card for unloading/loading of JSON instead of using this datastub. When doing the JSON part we might as well try and save the solved Databin with.

File: RPG Wizardry/Assets/Scripts/ResearchData/DataManager.cs (hunk 13 above)
Comment (Sanko-Kallig) [reply RPG Wizardry/Assets/Scripts/ResearchData/DataManager.cs]: Playerinventory was unnecessary in the Datamanager, removed it.

File: RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs (hunk 12 above)
Comment (Sanko-Kallig) [reply RPG Wizardry/Assets/Scripts/ResearchData/DataBin.cs]: Done!
